src/Http/WebUtilities/src/MultipartSectionPipeReader.cs

```diff
@@ -0,0 +1,366 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.WebUtilities
+{
+    internal class MultipartSectionPipeReader : PipeReader
+    {
+        private readonly PipeReader _pipeReader;
+        private bool _finished = false;
+        private bool _metadataSkipped = false;
+        private readonly MultipartBoundary _boundary;
+        private int _partialMatchIndex = 0;
+        private ReadOnlySequence<byte> _buffer;
+        private SequencePosition _endPosition;
+        private bool _isReaderComplete;
+
+        private static ReadOnlySpan<byte> CrlfDelimiter => new byte[] { (byte)'\r', (byte)'\n' };
+        private static ReadOnlySpan<byte> EndOfFileDelimiter => new byte[] { (byte)'-', (byte)'-' };
+
+        public long RawLength { get; private set; } = 0;
+
+        public long? LengthLimit { get; private set; }
+        // public long Length { get; private set; } = 0;
+
+        internal MultipartSectionPipeReader(PipeReader pipeReader, MultipartBoundary boundary)
+        {
+            _pipeReader = pipeReader;
+            _boundary = boundary;
+        }
+
+        private void UpdateLength(long read)
+        {
+            //Length += read;
+            RawLength += read;
+            if (LengthLimit.HasValue && RawLength > LengthLimit.GetValueOrDefault())
+            {
+                throw new InvalidDataException($"Multipart body length limit {LengthLimit.GetValueOrDefault()} exceeded.");
+            }
+        }
+
+        public override Stream AsStream(bool leaveOpen = false)
```

Comment:
This override isn't needed anymore, right? You can get rid of the custom PipeReaderStream and use the default?

Reply:
It won't work, the default version of PipeReaderStream throws an exception when calling it's Length.

Reply:
That's the expected behavior for a non-seekable stream. Why do you need length here? We just discussed how it was usually reported as 0 anyways.